### PR TITLE
feat: Add analyzer for ValueSourceAttribute

### LINKED ISSUE #93: Analyzer to ensure that the "target" of ValueSourceAttribute exists and is static
The target of `ValueSourceAttribute` should refer to a static field, property or method.

Much of the logic can probably be shared with #92.

In a subsequent task we can extend this with more of the requirements from https://github.com/nunit/docs/wiki/ValueSource-Attribute

For example
> If sourceType is specified, it represents the class that provides the data. **It must have a default constructor.**
> ...
> For the "target"
> * It must return an IEnumerable or a type that implements IEnumerable.
> * The individual items returned from the enumerator must be compatible with the type of the parameter on which the attribute appears.

### PR DESCRIPTION
**Pull request by mikkelbu**
Fixes #93

### CODE REVIEW

File: src/nunit.analyzers/ValueSourceUsage/ValueSourceUsageAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+                        if (method.Parameters.Length != 0)
+                        {
+                            context.ReportDiagnostic(Diagnostic.Create(
+                                methodExpectParameters,
+                                syntaxNode.GetLocation(),
+                                method.Parameters.Length));
+                        }
+                        break;
+                }
```
Comment (Dreamescaper) [inline src/nunit.analyzers/ValueSourceUsage/ValueSourceUsageAnalyzer.cs:153]: (feel free to ignore)

```suggestion                
                var memberType = symbol switch
                {
                    IPropertySymbol property => property.Type,
                    IFieldSymbol field => field.Type,
                    IMethodSymbol method => method.ReturnType,
                    _ => null
                };

                if (memberType != null)
                {
                    ReportIfSymbolNotIEnumerable(context, syntaxNode, memberType);
                }

                if(symbol is IMethodSymbol methodSymbol && methodSymbol.Parameters.Length != 0)
                {
                    context.ReportDiagnostic(Diagnostic.Create(
                        methodExpectParameters,
                        syntaxNode.GetLocation(),
                        methodSymbol.Parameters.Length));
                }
```

File: src/nunit.analyzers/ValueSourceUsage/ValueSourceUsageAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                        { SourceCommonConstants.PropertyKeyNameOfTarget, nameOfTarget }
+                    };
+
+                    context.ReportDiagnostic(Diagnostic.Create(
+                        considerNameOfDescriptor,
+                        syntaxNode.GetLocation(),
+                        properties.ToImmutableDictionary(),
+                        nameOfTarget,
+                        stringConstant));
+                }
```
Comment (Dreamescaper) [inline src/nunit.analyzers/ValueSourceUsage/ValueSourceUsageAnalyzer.cs:124]: Could you extract that to separate method? E.g. `ReportToUseNameOfIfApplicable`.

Quite a lot happens in this method, a bit hard to keep track

File: src/nunit.analyzers/ValueSourceUsage/ValueSourceUsageAnalyzer.cs (hunk 1 above)
Comment (Dreamescaper) [reply src/nunit.analyzers/ValueSourceUsage/ValueSourceUsageAnalyzer.cs:153]: Additional suggestion - to move the `memberType` switch to `ReportIfSymbolNotIEnumerable` method - to make `AnalyzeAttribute` a bit more concise.

Comment (mikkelbu) [reply src/nunit.analyzers/ValueSourceUsage/ValueSourceUsageAnalyzer.cs:153]: I actually tried to something similar, but it did not occur to me that I could use switch expressions (don't think that I've ever use them). I like the change 👍

File: src/nunit.analyzers/ValueSourceUsage/ValueSourceUsageAnalyzer.cs (hunk 2 above)
Comment (mikkelbu) [reply src/nunit.analyzers/ValueSourceUsage/ValueSourceUsageAnalyzer.cs]: Sounds like a good suggestion, then we can also use it for the `TestCaseSourceUsesStringAnalyzer`

**Pull request merged**